strikt-arrow/src/main/kotlin/strikt/arrow/either/Either.kt

```diff
@@ -0,0 +1,99 @@
+package strikt.arrow.either
+
+import arrow.core.Either
+import arrow.core.Predicate
+import strikt.api.Assertion
+import strikt.api.expectThat
+
+@Suppress("UNCHECKED_CAST")
+fun <L, R> Assertion.Builder<Either<L, R>>.isRight() =
+  assert("should be Right") {
+    when (it) {
+      is Either.Right -> pass()
+      is Either.Left -> fail()
+    }
+  } as Assertion.Builder<Either.Right<R>>
+
+@Suppress("UNCHECKED_CAST")
+fun <L, R> Assertion.Builder<Either<L, R>>.isRight(value: R) =
+  assert("should be Right($value)") {
+    when (it) {
+      is Either.Right -> if (it.b == value) {
+        pass()
+      } else {
+        fail()
+      }
+
+      else -> fail()
+    }
+  } as Assertion.Builder<Either.Right<R>>
+
+@Suppress("UNCHECKED_CAST")
+fun <L, R> Assertion.Builder<Either<L, R>>.isRightWhere(predicate: Predicate<R>) =
```

Comment:
IMO I think it is better to reduce the surface area and amount of assertions and predicates here so that it is better aligned with composing other assertions built into the `core` library.

For example, instead of 

```kotlin
fun <L, R> Assertion.Builder<Either<L, R>>.isRightWhere(predicate: Predicate<R>)
```

It would be better for a user to do something like:

```kotlin
expectThat(myEither)
  .isRight()
  .b
   .isEqualTo(value)
```

Or even have `isRight()` unpack the wrapped value (maybe?)

```kotlin
expectThat(myEither)
  .isRight()
  .b
  .isEqualTo(value)
```

Where `b` is

```kotlin
val <R> Builder<Either.Right<R>>.b: Builder<B>
  get() = get("right value") { b }
```

For the `Predicate` case above, a user would use the core built in assertions 

```kotlin
expectThat(myEither)
  .isRight()
  .and {
    // Assertion.Builder<R> where <R> is type of `Either.Right.b`
  }
```

WDYT?

Reply:
i agree, i think thats a good change.

i removed the functions with the predicate and introduced extension properties to allow traversal of the unwrapped values.

i did not remove the function `isRight/Left(value)`. The check for value equality seems to be a very common case and its a bit shorter to write `expectThat(anEither).isLeft("left")` instead of `expectThat(anEither).isLeft().a.isEqualTo("left")`.

Is it possible to change the assertion subject within a custom assertion?
I tried to let `isRight()` do the unwrapping of the value so the return type is `Builder<R>` instead of `Builder<Either.Right<R>>`.  however i did not find a way yet.

right now you still have to write:
```
expectThat(myEither)
   .isRight()
   .b
   .and {
        ...
   }
```

Reply:
It is possible. 
Maybe [this](https://github.com/sandjelkovic/kombinator/blob/0adc5ea5f43406308662b2347c8d8c70b00b6d23/kombinator-spring-mvc/src/test/kotlin/com/sandjelkovic/kombinator/test/StriktExtensions.kt#L30) will help, it's an example of automatic unwrap within and after isRight/isLeft assertions (via a lambda block)

Reply:
> i did not remove the function isRight/Left(value). The check for value equality seems to be a very common case and its a bit shorter to write 

I could see value in both removing it and keeping it, so I'd defer the decision to whatever others think is best

------

Yeah, maybe the chained approach described above might not work:

```kotlin
expectThat(myEither)
   .isRight()
   .b
```

That might run into complications with typing and chaining and I'm not sure at the moment if assertion chains continue execution if an earlier assertion in the chain has failed. So, would `.b` bomb out if the value is an `Either.Left`, or would those assertions not be evaluated (or skipped)?

Might be related to https://github.com/robfletcher/strikt/issues/8 ?

Any thoughts @robfletcher ?

------

@sandjelkovic I like your linked approach, that feels like a good way, too.

Reply:
thanks @sandjelkovic. 

i put both variants next to each other to see how they differ when used (_isRight2_ is the version from @sandjelkovic)

```
 test("can assert on type") {
      expectThat(anEither).isRight()
      expectThat(anEither).isRight2()
    }

    test("can assert on type and value equality") {
      expectThat(anEither).isRight("right")
      expectThat(anEither).isRight2 { isEqualTo("right") }
    }

    test("can assert on type and traverse unwrapped value") {
      expectThat(anEither).isRight().b.isEqualTo("right")
      expectThat(anEither).isRight2().isEqualTo("right")
    }

    test("can have nested assertions on unwrapped value") {
      expectThat(Either.right(MyTuple("myName", 1, "uuid"))).isRight().b.and {
        get { name }.isEqualTo("myName")
        get { id }.isNotNull().isGreaterThan(0L)
        get { uuid }.isNotNull().isNotBlank()
      }

      expectThat(Either.right(MyTuple("myName", 1, "uuid"))).isRight2().and {
        get { name }.isEqualTo("myName")
        get { id }.isNotNull().isGreaterThan(0L)
        get { uuid }.isNotNull().isNotBlank()
      }
    }
``` 

I also like this approach as it gets rid of the unwrap step. on the other hand i consider this  behavior a bit surprising as i could not find other parts in strikt that do implicit subject traversal.

Reply:
Yeah, that is another one I could see an argument for either side. I'm alright with either approach.